### ISSUE TITLE
Explain Mercator transformation and dimensions

### ISSUE THREAD

**Issue by mlshapiro** [creare-com/podpac#289]
**Description**

When we transform from EPSG:4326 (lat, lon) to EPSG:3857 (Mercator, x, y) the "lat" label stores "x" values and the "lon" label stores the "y" values.  Intuitively I see the "x" being "lon" values and the "y" the "lat" values.

@mpu-creare also noticed the transformed coordinates change shape.

**Steps to Reproduce**
```
c = Coordinates([clinspace(43, 40, 5), clinspace(-76, -72, 2)], dims=['lat', 'lon'])
c.shape 
>> (5, 2)
cm = c.transform('epsg:3857')
cm.shape
>> (2, 5)
```

**Expected Behavior**
```
c = Coordinates([clinspace(43, 40, 5), clinspace(-76, -72, 2)], dims=['lat', 'lon'])
c.shape 
>> (5, 2)
cm = c.transform('epsg:3857')  # lat = y, lon = x
cm.shape
>> (5,2)  
```

cm['lat'] == "y" values in mercator dimensions
cm['lon'] == "x" values in mercator dimensions

**Comment by jmilloy** [comment]
I think shape issue is here:

```
                c = DependentCoordinates(
                    np.meshgrid(c1.coordinates, c2.coordinates),
                    dims=[c1.name, c2.name],
                    ctypes=[c1.ctype, c2.ctype],
                    segment_lengths=[c1.segment_lengths, c2.segment_lengths])
```

It should possibly be something like this:

```
                coords1, coords2 = np.meshgrid(c1.coordinates, c2.coordinates)
                c = DependentCoordinates(
                    [coords1.T, coords2.T],
                    dims=[c1.name, c2.name],
                    ctypes=[c1.ctype, c2.ctype],
                    segment_lengths=[c1.segment_lengths, c2.segment_lengths])
```

**Comment by mpu-creare** [comment]
Perhaps we want:
``` 
np.meshgrid(..., indexing='ij')
```
and perhaps the order of the inputs also need reversing... ?

**Comment by jmilloy** [comment]
Okay, `indexing='ij'` works to fix the shape, too.

```
                c = DependentCoordinates(
                    np.meshgrid(c1.coords, c2.coords, indexing=ij),
                    dims=[c1.name, c2.name],
                    ctypes=[c1.ctype, c2.ctype],
                    segment_lengths=[c1.segment_lengths, c2.segment_lengths])
```

**Comment by jmilloy** [comment]
> and perhaps the order of the inputs also need reversing... ?

I think the issue is that *some* transformations may need to switch the order, but not all. The relevant code is in `DependentCoordinates._transform`:

```
            ilat = self.dims.index('lat')
            ilon = self.dims.index('lon')

            # coordinates
            lat = coords[ilat].flatten()
            lon = coords[ilon].flatten()
            tlat, tlon = transformer.transform(lat, lon)
            coords[ilat] = tlat.reshape(self.shape)
            coords[ilon] = tlon.reshape(self.shape)
```

When transform 4326 to 3857, really `tlat` is mercator x-values and `tlon` is mercator y-values, and would be more intuitively be swapped. If you swap them, then the transformed coordinates indeed match intuition and use `'lat'` for y-values and `'lon'` for x-values, but then transforming back gets messed up because you pass the coordinates to in the wrong order to `transformer.transform`.

I feel like this can only be fixed with a list of which projections should be swapped. When transforming from a "normal" crs to a "swapped" crs, you swap the output (tlon and tlat). When transforming from a "swapped" crs to a "normal" crs, you swap the input (lat and lon). Otherwise you leave it alone.

I guess there could be a way to detect which ones should be swapped automatically? Mainly it sounds like this is an issue when transforming from lat-lon coordinates to x-y coordinates.

**Comment by jmilloy** [comment]
So yeah, we can inspect the `pyproj.CRS` object and deal with it automatically a few ways by looking at the Axis Info.
 * I think ellipsoid is what I called "normal" above, and cartesian is "swapped". I'm not sure if that generalizes to other projections or not, but I would bet so.
 * Alternatively, if it uses Lat and Lon we call it "normal", X and Y we call it "swapped", and anything else may need a different case/behavior.
 * Basically we want "[north]" coordinates to be lat and "[east]" coordinates to be lon, and that info is included as well.

We could also generalize our coordinates so that they natively handle cartesian coordinates. Currently they only properly handle ellipsoid coordinates, because lat and lon are the only spatial dimensions allowed. We could add x and y as allowed dimensions and handle transforming between them. This is mostly a cosmetic change, and some validation to require the dimension names to match the CRS.

Estimates:
 - list of "swapped" CRS: 1 hour
 - detect "swapped" (cartesian) CRS automatically: 2 hours
 - add x and y dimensions to coordinates: 16 hours

**Comment by mlshapiro** [comment]
Great - I will implement the shape fix:

```
...
indexing='ij'
```

As for the ordering, its all right there in the `PROJ4` documentation: https://proj.org/faq.html#why-is-the-axis-ordering-in-proj-not-consistent . I need to read through this and then I'll comment on the implementation

**Comment by mlshapiro** [comment]
The relevant section may be: 

> In most cases coordinate reference systems with geodetic coordinates expect the input ordered as latitude/longitude (typically with the EPSG dataset), however, internally PROJ expects an longitude/latitude ordering for all projections. This is generally hidden for users but in a few cases it is exposed at the surface level of PROJ, most prominently in the proj utility which expects longitude/latitude ordering of input date (unless proj -r is used).

**Comment by mlshapiro** [comment]
I think this can be fixed pretty simply using the from_proj parameter `always_xy`. Working now.

**Comment by jmilloy** [comment]
Very cool, pyproj is a good library.

**Comment by mlshapiro** [comment]
This worked, pushing to a branch - @jxm would you be willing to review quickly? 

This will have some unforseen consequences for datalibs that used x/y backwards. Noteable we need to reverse EGI

**Comment by jmilloy** [comment]
Yeah, go ahead and assign me the PR when it is ready. Do you have time to include a test for the always_xy change and a test for the shape issue that you already pushed? Let me know if you want me to do those tests.

**Comment by mlshapiro** [comment]
I will have time to fix the tests in an hour

**Comment by mlshapiro** [comment]
Fixed in #290
